_posts/2020-12-04-release-1.12.0.md

```diff
@@ -0,0 +1,313 @@
+
+**Watermark Pushdown in the Kafka Connector ([FLINK-20041](https://issues.apache.org/jira/browse/FLINK-20041))**
+
+To ensure correctness when consuming from Kafka, it’s generally preferable to generate watermarks on a per-partition basis, since the out-of-orderness within a partition is usually lower than across all partitions. Flink will now push down watermark strategies to [emit **per-partition watermarks**]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/table/connectors/kafka.html#source-per-partition-watermarks) from within the Kafka consumer. The output watermark of the source will be determined by the minimum watermark across the partitions it reads, leading to better (i.e. closer to real-time) watermarking. Watermark pushdown also lets you configure per-partition **idleness detection** to prevent idle partitions from holding back the event time progress of the entire application.
+
+**Newly Supported Formats**
+
+<table class="table table-bordered" style="font-size:95%">
+  <thead>
+    <tr>
+      <th>Format</th>
+      <th>Description</th>
+      <th>Supported Connectors</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td><a href="https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/connectors/formats/avro-confluent.html">Avro Schema Registry</a></td>
+      <td>Read and write data serialized with the Confluent Schema Registry <a href="https://docs.confluent.io/platform/current/schema-registry/serdes-develop/serdes-avro.html">KafkaAvroSerializer</a>.</td>
+      <td>Kafka, Upsert Kafka</td>
+    </tr>
+    <tr>
+      <td><a href="https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/connectors/formats/debezium.html">Debezium Avro</a></td>
+        <td>Read and write Debezium records serialized with the Confluent Schema Registry KafkaAvroSerializer.</td>
+        <td>Kafka</td>
+    </tr>
+      <td><a href="https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/connectors/formats/raw.html">Raw</a></td>
+      <td>Read and write raw (byte-based) values as a single column.</td>
+      <td>
+        <p>Kafka, Upsert Kafka</p>
+        <p>Kinesis</p>
+        <p>FileSystem</p>
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+**Multi-input Operator for Join Optimization ([FLINK-19621](https://issues.apache.org/jira/browse/FLINK-19621))**
+
+Shuffling is one of the most expensive operations in a Flink job. To eliminate unnecessary shuffles and improve the performance of batch and streaming Table API/SQL jobs, the default planner now leverages the N-ary stream operator introduced in the last release ([FLIP-92](https://cwiki.apache.org/confluence/display/FLINK/FLIP-92%3A+Add+N-Ary+Stream+Operator+in+Flink)) to implement the "chaining" of operators connected by forward edges and change these network shuffles to local function calls.
```

Comment:
This is not really reducing shuffles, but mainly serialization overhead.
Maybe a good way to describe this is supporting to fuse N-way joins into a single task (where they previously were a cascade of 2-way joins) and this eliminate the overhead for serialization and thread-handover between tasks.

Reply:
Is it incorrect? This is aligned with the [design document](https://docs.google.com/document/d/1qKVohV12qn-bM51cBZ8Hcgp31ntwClxjoiNBUOqVHsI/edit#heading=h.8lnz38suktxb), which seems to put a lot of emphasis on eliminating redundant shuffles.

Reply:
I think there is unfortunately a lot of misleading terms in the design doc.

> These forward shuffles, however, can be eliminated.

To my mind, there is no such thing as a _"forward shuffle"_. The name shuffle implies that there is data redistribution. The "forward" connection in Flink is passing data from one task to one other task.

What we previously had in these places is a 1:1 data exchange between tasks. That always involved serialization.
Sometimes this exchange was a "batch exchange" that spilled data to disk, to decouple when the task need to be scheduled. But none typically involved any network (as a shuffle does). Granted, for the batch cases, saving the disk I/O is actually even a bigger deal than the serialization.

But a real shuffle was never saved. I imagine also that users would find it puzzling to understand how fusing more operators together all of a sudden removes the need to transport data over the network (shuffle data).